flows/engine/assets.go

```diff
@@ -11,95 +13,165 @@ import (
 	"github.com/nyaruka/goflow/utils"
 )
 
-type assetContainer struct {
-	assetType  flows.AssetType
-	asset      flows.Asset
-	accessedOn *time.Time
-	expiresOn  time.Time
-	fetchURL   string
+type assetURL string
+type itemUUID string
+
+type assetType string
+
+const (
+	assetTypeObject assetType = "object"
+	assetTypeSet    assetType = "set"
+)
+
+type assetItemType string
+
+const (
+	assetItemTypeChannel assetItemType = "channel"
+	assetItemTypeFlow    assetItemType = "flow"
+	assetItemTypeGroup   assetItemType = "group"
+)
+
+// container for any asset in the cache
+type cachedAsset struct {
+	asset      interface{}
+	addedOn    time.Time
+	accessedOn time.Time
 }
 
-type assetStore struct {
-	cache      map[flows.AssetUUID]assetContainer
-	cacheMutex sync.Mutex
+// AssetCache fetches and caches assets for the engine
+type AssetCache struct {
+	cache map[assetURL]cachedAsset
+	mutex sync.Mutex
 }
 
-func NewAssetStore() flows.AssetStore {
-	return &assetStore{cache: make(map[flows.AssetUUID]assetContainer)}
+// NewAssetCache creates a new asset cache
+func NewAssetCache() *AssetCache {
+	return &AssetCache{cache: make(map[assetURL]cachedAsset)}
 }
 
-func (m *assetStore) requestAsset(uuid flows.AssetUUID, assetType flows.AssetType) (flows.Asset, error) {
-	m.cacheMutex.Lock()
-	defer m.cacheMutex.Unlock()
+func (c *AssetCache) putAsset(url assetURL, asset interface{}) {
+	c.cache[url] = cachedAsset{asset: asset, addedOn: time.Now().UTC()}
+}
+
+func (c *AssetCache) addAsset(url assetURL, asset interface{}) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.putAsset(url, asset)
+}
 
-	container, found := m.cache[uuid]
+// gets an asset from the cache if it's there or from the asset server
+func (c *AssetCache) getAsset(url assetURL, aType assetType, itemType assetItemType) (interface{}, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	cached, found := c.cache[url]
 	if !found {
-		return nil, fmt.Errorf("no such asset of type '%s' with UUID '%s'", assetType, uuid)
-	}
+		asset, err := c.fetchAsset(url, aType, itemType)
+		if err != nil {
+			return nil, err
+		}
 
-	if container.asset == nil || time.Now().After(container.expiresOn) {
-		// TODO try to (re)fetch from URL, and validate
+		c.putAsset(url, asset)
+		return asset, nil
 	}
 
-	accessedOn := time.Now().UTC()
-	container.accessedOn = &accessedOn
+	// update the accessed on time
+	cached.accessedOn = time.Now().UTC()
 
-	return container.asset, nil
+	return cached.asset, nil
 }
 
-func (m *assetStore) AddAsset(asset flows.Asset, fetchURL string) {
-	m.cache[asset.AssetUUID()] = assetContainer{
-		assetType: asset.AssetType(),
-		asset:     asset,
-		expiresOn: time.Now().Add(5 * time.Minute),
-		fetchURL:  fetchURL,
+// fetches an asset by its URL and parses it as the provided type
+func (c *AssetCache) fetchAsset(url assetURL, aType assetType, itemType assetItemType) (interface{}, error) {
+	response, err := http.Get(string(url))
```

Comment:
What about authentication for these fetches? Is the expectation that none of these are gated and not publicly addressable as a requirement?

Reply:
If we're using the same endpoint for the flow editor then we need some authentication. I was figuring we'd support regular session auth for the flow editor and maybe a token based system for the flow server?